google-cloud-spanner/src/main/java/com/google/cloud/spanner/XGoogSpannerRequestId.java

```diff
@@ -81,8 +130,56 @@ public boolean equals(Object other) {
         && Objects.equals(this.attempt, otherReqId.attempt);
   }
 
+  public void incrementAttempt() {
+    this.attempt++;
+  }
+
+  Map<SpannerRpc.Option, ?> withOptions(Map<SpannerRpc.Option, ?> options) {
+    Map copyOptions = new HashMap<>();
+    if (options != null) {
+      copyOptions.putAll(options);
+    }
+    copyOptions.put(SpannerRpc.Option.REQUEST_ID, this);
+    return copyOptions;
+  }
+
   @Override
   public int hashCode() {
     return Objects.hash(this.nthClientId, this.nthChannelId, this.nthRequest, this.attempt);
   }
+
+  interface RequestIdCreator {
+    XGoogSpannerRequestId nextRequestId(long channelId, int attempt);
+  }
+
+  static class NoopRequestIdCreator implements RequestIdCreator {
+    NoopRequestIdCreator() {}
+
+    @Override
+    public XGoogSpannerRequestId nextRequestId(long channelId, int attempt) {
+      return XGoogSpannerRequestId.of(1, 1, 1, 0);
+    }
+  }
+
+  static void assertMonotonicityOfIds(String prefix, List<XGoogSpannerRequestId> reqIds) {
```

Comment:
I would rather expect this code in a test file than here

Reply:
Done and thanks!